### PR TITLE
Add attachment descriptions

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

This implements attachment descriptions; they are already available, but gated behind an experiment in clients (`2021-09_inline_attachment_uploads`).

There is a change in the attachment handling, which would've generally resulted in previous attachments not being kept by default on edit anymore, *if* new files are also being uploaded.
This would affect types that don't contain information about existing attachments, i.e. all types except `Message` and `InteractionMessage`.
To fix this, the original message will be fetched for those types and existing attachments will carry over from that, if new files are being added and the developer didn't explicitly provide the `attachments` kwarg.

Providing existing *and* new attachments in the `attachments` list like this when new files are being uploaded will be required for API v10 at the latest. API v8 itself doesn't require it yet, however it is already needed when using attachment descriptions.

### TODO
- ~~[ ] Check if `attachment://` filenames in embeds keep working as intended~~

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by shiftinv** [comment]
One possible solution to fix the issue for the other types noted above and to avoid those breaking changes:
Have the library fetch the message if required, and carry over the attachments by default, which is how it works for `Message` and `InteractionMessage` as well, albeit without having to fetch the message.
This would obviously introduce an additional API request, but only in those cases where new files are being added on edit, which might be acceptable?

**Comment by EQUENOS** [comment]
> This would obviously introduce an additional API request, but only in those cases where new files are being added on edit, which might be acceptable?

I think this is acceptable

**Comment by shiftinv** [comment]
Alright, I rewrote it a bit, no more breaking changes as far as I'm concerned. I'm fairly happy with the PR now, and updated the main description above to reflect the new behavior, which boils down to "fetch the original message and take its attachments if required".
Still need to take a look at the `attachment://` stuff though.

**Comment by thesadru** [comment]
Oh no, did I fuck something up with those?

**Comment by shiftinv** [comment]
> Oh no, did I fuck something up with those?

No, it's all good, I'm just not sure if/how these changes interact with the `attachment://` filenames, since there seems to be some weirdness in Discord's API with those and the `attachments` parameter.

**Comment by shiftinv** [comment]
Right, that's the weird behavior I was afraid would come up:
  - Send an embed with `url=...` - the response contains the embed and `attachments=[]`
    - When you edit the message and set `attachments=[]`, nothing changes
  - Send an embed with `file=...` - the response also contains the embed and `attachments=[]`
    - When you edit the message and set `attachments=[]`, the image in the embed disappears

I'm guessing files used with the `attachment://` urls are still part of the `attachments` list internally, but the API hides them if they're referenced in an embed. You still have to somehow provide them when setting `attachments=` yourself though.

Can't really think of a way to solve this... extracting the attachment IDs from the embed's image urls and adding them to the `attachments` list would mostly work, but there's no way of knowing if an embed image url is actually referencing an attachment of the message or something else (in which case you'd get `Attachment data not found`) :/

**Comment by shiftinv** [comment]
tl;dr: Adding new files to a message that has embeds with `attachment://` image urls will cause the images in the embed to disappear.

Maybe just keeping it like this and adding a note to the docs would be fine.

**Comment by EQUENOS** [comment]
Yeah I think we can keep it like this, because the case you described is very specific.

**Comment by thesadru** [comment]
I'm a bit confused. Does that mean a paginator with embeds which only use `file=` are not possible?

**Comment by shiftinv** [comment]
> I'm a bit confused. Does that mean a paginator with embeds which only use `file=` are not possible?

That should still work fine, I think you'd want to remove the existing image anyway. If you were to replace an existing embed that uses `file=` with a new one on the current master branch and don't set `attachments=[]`, the previous embed's image will pop out and display separately.
